### ISSUE TITLE
Some malformed inputs raise TypeError instead of ValidationError

### ISSUE THREAD

**Issue by jameysharp** [pydantic/pydantic#2762]
Since this is my second weird bug report in two days, let me first say that Pydantic has enabled me to prototype big chunks of a somewhat complex protocol very quickly, and I really appreciate what you've built. I feel I'm just stumbling around into odd corners as I learn to use this tool.

### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.8.1
            pydantic compiled: False
                 install path: /nix/store/1f6nk37c7df54cm8b3rlnn6pidaa3mrp-python3.8-pydantic-1.8.1/lib/python3.8/site-packages/pydantic
               python version: 3.8.8 (default, Feb 19 2021, 11:04:50)  [GCC 9.3.0]
                     platform: Linux-5.4.114-x86_64-with
     optional deps. installed: ['typing-extensions']
```

The following call raises `TypeError` when I expected it to raise `ValidationError`:

```py
import pydantic

pydantic.BaseModel.parse_obj([[None, None]])
```

This came up when throwing random JSON at my API using Hypothesis.

`parse_obj` looks like this in the Pydantic version I'm using:

```py
    @classmethod
    def parse_obj(cls: Type['Model'], obj: Any) -> 'Model':
        obj = cls._enforce_dict_if_root(obj)
        if not isinstance(obj, dict):
            try:
                obj = dict(obj)
            except (TypeError, ValueError) as e:
                exc = TypeError(f'{cls.__name__} expected dict not {obj.__class__.__name__}')
                raise ValidationError([ErrorWrapper(exc, loc=ROOT_KEY)], cls) from e
        return cls(**obj)
```

I guess what's happening is: `obj` is not a `dict`, but it _is_ an iterable of pairs, so it converts to a `dict` without trouble; however, the `cls(**obj)` call requires that the dictionary's keys are all strings, which is not checked.

This doesn't seem like a security bug to me since Pydantic doesn't allow the invalid data through either way, but I do feel like it's a bug that under some circumstances the exception that's thrown is `TypeError` instead of `ValidationError`.

I'm also surprised to learn that at the root of an object, Pydantic treats iterables of pairs the same as it treats dictionaries. I understand the [Data Conversion](https://pydantic-docs.helpmanual.io/usage/models/#data-conversion) warning in the documentation, I think, but I didn't know that extended beyond fields. In fact the documentation for `parse_obj` says it shouldn't:

> parse_obj: this is very similar to the `__init__` method of the model, except it takes a dict rather than keyword arguments. If the object passed is not a dict a `ValidationError` will be raised.

I'm now working around it by explicitly checking that I got a `dict` from JSON decoding before passing it to Pydantic, which I guess is fine; I just feel like this needs stronger warnings.

**Comment by samuelcolvin** [comment]
Thanks for reporting, this is definitely a bug, I guess it might be partially related to #1268.

But it can also be produced with a real `dict`:

```py
from pydantic import BaseModel


class Foobar(BaseModel):
    foo: str


Foobar.parse_obj({1: 2})
```

i guess until v2 we should just catch this `TypeError` and raise a `__root__` validation error.

I'm really surprised this has just been reported, I wonder if it's a regression and at some point in the past was caught.

**Comment by jameysharp** [comment]
Ooh, right, I didn't think about dictionaries with non-string keys. Though I think my workaround is still valid for me since the only data I'm handing to Pydantic comes from `json.loads`, which should never produce such a dictionary.

I just tried to write a Hypothesis test for Pydantic's test suite like the following, but couldn't come up with a general strategy I could pass to `@given` that would hit the interesting cases:

```py
class Never(pydantic.BaseModel, extra="forbid"):
    @pydantic.root_validator
    def never(cls, values):
        raise ValueError("refusing (presumably empty) input")

@given(...)
def test_always_validation_error(obj):
    with pytest.raises(pydantic.ValidationError):
        Never.parse_obj(obj)
```

I thought the example in [`hypothesis.strategies.from_type`](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.from_type) of "generate all known types" would be nice (`from_type(type).flatmap(from_type)`) but it didn't really explore interesting parts of the input domain.

**Comment by jameysharp** [comment]
I think #1268 is only superficially related, since as you pointed out even requiring the input to be a `dict` isn't sufficient. This seems like a separate issue to me. So I think even after v2 it's still necessary to do as you suggested: catch the `TypeError` and raise a `__root__` validation error.

I've just discovered that it isn't even sufficient to require that the dictionary have only `str` keys, because `parse_obj({"self": 0})` raises `TypeError` as well.

As far as I can tell, `parse_obj({0: 0})` has raised `TypeError` ever since `parse_obj` was introduced in 044fd42f4fda6b06bedbe5f42fc8324020706e27. It's additionally done so on `parse_obj([[0, 0]])` ever since 65d717a7b7b09e43074e575c7694980d1dda4af1 when you first made `parse_obj` accept dict-like objects to fix #472.

Given only valid JSON input, the former case can only be triggered with `{"self": ...}`, which I imagine would make it hard to stumble across in the usual situations where people use Pydantic, but the latter can be triggered with a pretty wide range of inputs.

I used this program (or the equivalent with the list of pairs) to quickly test specific commits with e.g. `git bisect run`:

```py
import pydantic, pytest
with pytest.raises(pydantic.ValidationError):
    pydantic.BaseModel.parse_obj({0: 0})
```

I'm still thinking about exactly which inputs are affected by this issue, so:

I just realized the same error can be triggered with e.g. `parse_raw('[[null, null]]')` and presumably with `parse_file` as well. I don't think anyone would expect those methods to support "dictionary-like" inputs.

On the other hand, I've checked that passing a list of pairs to a nested model turns the `TypeError` into a `ValidationError`, so this is only an issue for the top-level model. The validation error message isn't ideal, but that will be fixed by fixing this issue.

```py
from pydantic import BaseModel
class Foo(BaseModel):
    bar: BaseModel
Foo.parse_obj({"bar": [[None, None]]})
```
```
pydantic.error_wrappers.ValidationError: 1 validation error for Foo
bar
  __init__() keywords must be strings (type=type_error)
```
